### PR TITLE
Network: Retry requests that failed in flight

### PR DESCRIPTION
**Pull request by kidroca**
### Details

Add updates to:
- Retry requests when something unexpected caused an error in flight
- Prevent multiple authentication calls running at the same time

### Fixed Issues

$ https://github.com/Expensify/App/issues/5952

### Tests

We're trying to make a request fail in the middle, like connection loss
The easiest way so far has been using Chrome to "slow down the connection" 
On staging it's even enough to cause a timeout (but on localhost I think there are different rules and it doesn't happen)
Check out the web before after video for a thorough demonstration 

#### Web (Chrome) & Desktop
1. Open dev tools and add a very slow connection profile
2. Enable the throttled connection you just made
3. Write a message or two 
  - on staging the request should fail in about 15 sec. just wait
  - on localhost you can kill your router connection now to cause the request to fail
      - you'll probably need to also kill the proxy server and restart it
      - also you might need to go to offline and online (in Chrome) to let it know it really don't have internet..
4. Verify the messages failed (see `Report_AddComment`) in the network tab
5. Set the Network speed to "offline" - now we're acting like we're fully offline
6. Send a few more messages, but do take a note where the message from step 3) are
7. Now set Network speed to "full" and observe the requests getting retried
  - if you tried this on localhost resume you're real internet connection before setting network speed to full in Chrome
8. There should be no pending messages stuck on the bottom 

#### Others
We're trying to do the same as above, but without dev helpers
It sometimes possible to trick the device you're still online, like if you unplug your network cable or use a tethered connection and disable "data" on the tethered connection, then the requests should timeout and fail
When you resume connectivity any pending request should be sent

### QA Steps

See if you can do the "Tests" above, it should be easier on staging and Chrome, just letting the connections timeout using dev tools and throttle
Otherwise a real world case it to use a device in a spotty connection and notice that you're getting offline or messages are sent too slow
Check for regressions around network related stuff like uploading an attachment
Spending some time in airplane mode then resuming
Use the app for a while and monitor for any indication of a regression that might be caused here

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

##### Before and After (after starts around the 20th second mark)
https://user-images.githubusercontent.com/12156624/144518079-e8face7f-f0dd-44b5-8a2f-950e0828bbc4.mp4

#### Mobile Web

image

#### Desktop

image

#### iOS

image

#### Android

https://user-images.githubusercontent.com/12156624/144518247-b11bdb5f-6174-413f-892b-04bf6c57cbe3.mp4

### CODE REVIEW

File: src/libs/API.js
Comment (kidroca) [inline src/libs/API.js:120]: Pausing the queue and setting the `isAuthenticating` flag is moved to `Authenticate` - when `Authenticate` is called it would raise the flag, and set it back to `false` once it's done

Calling `addDefaultValuesToParameters` is not necessary `Network.post` already does it

`if (isAuthenticating)`, then and catch are no longer necessary - we call reauthenticate and finally retry the original request
I used `finally` since the original logic retries the request on failures as well. This is so because it probably triggers re-authenticate again...

File: src/libs/actions/Session/index.js
Comment (kidroca) [inline src/libs/actions/Session/index.js:449]: I noticed this was still causing extra Push Authenticate calls 

Multiple Authenticate and Push Authenticate requests

File: tests/unit/NetworkTest.js
Comment (kidroca) [inline tests/unit/NetworkTest.js:188]: The order of the requests changed

You can see the original code expected request to be resolved in reverse order 
1. TEST_PERSONAL_DETAILS
2. TEST_ACCOUNT_DATA
3. TEST_CHAT_LIST

While the actual `API.GET` calls are
1. TEST_CHAT_LIST
2. TEST_PERSONAL_DETAILS
3. TEST_ACCOUNT_DATA

When we resume connectivity order of the requests shouldn't matter - they are quickly sent in parallel anyway
If for some reason we want to test the order then better to happen in a separate test

BTW the order now is preserved to match the `API.GET` original call order (second list above)

File: tests/unit/NetworkTest.js
Comment (kidroca) [inline tests/unit/NetworkTest.js:150]: It seems unnecessary to merge results to Onyx only to update the local variables here when we can update them directly

**Review by kidroca — COMMENTED** [review]
Added some inline notes on the changes

**Comment by kidroca** [comment]
Bump @johnmlee101 
This one have been ready for a while as well

File: src/libs/API.js
Comment (johnmlee101) [inline src/libs/API.js:79]: The function name `retry` is a bit odd considering the only thing that infers that its for retrying is the context in which its called. The code itself is just a basic post, right?

**Review by johnmlee101 — CHANGES_REQUESTED** [review]
One minor comment, rest seems good and I'll get another reviewer to double check

**Comment by johnmlee101** [comment]
cc @marcaaron @roryabraham if you want to take a crack at this as well and review in addition to me

**Comment by marcaaron** [comment]
Ah sorry I don't have time to deep dive into the networking code atm. I mentioned this somewhere else, but I think it would be valuable to improve the existing networking logic with a refactor of some kind before complicating it further with whatever changes we are proposing (it will take me some time to just catch up on what it is we're doing and why).

File: src/libs/API.js
Comment (kidroca) [reply src/libs/API.js:79]: it uses an existing request and re-posts it. also it does hook to the `then` and `catch` of the original

But yeah it's primarily to reuse the logic and spare some duplicate post code

**Comment by kidroca** [comment]
This PR untangles some stuff and actually  simplifies the netwotking code

Half of the changes are from the added unit tests 

A hold on the network issues, would block me on the on persisted file handling as well

File: src/libs/API.js
Comment (kidroca) [reply src/libs/API.js:79]: If it's only about the variable name I thought of naming it `retryRequest` I don't know why I didn't go with that...

File: src/libs/API.js
Comment (roryabraham) [inline src/libs/API.js:164]: Should we pause the request queue right here rather than in `Authenticate`? Because we know on line 55 if we're going to try to reauthenticate, so maybe we should immediate pause the request queue so that another request doesn't begin processing before we attempt to reauthenticate?

File: src/libs/API.js
Comment (marcaaron) [reply src/libs/API.js:120]: "queue pausing" is a concept that I'd recommend we get rid of entirely. It's confusing and basically "code" for whether we are authenticating or not.

File: src/libs/API.js
Comment (marcaaron) [inline src/libs/API.js:159]: We usually use `should*` for booleans and not functions.

File: src/libs/actions/Session/index.js
Comment (marcaaron) [reply src/libs/actions/Session/index.js:449]: I don't know what you think you are looking at. But anytime we subscribe to a channel a call to `Push_Authenticate` is made.

Comment (kidroca) [reply src/libs/actions/Session/index.js:449]: This opening/focusing App after more than 2 hours are passed since last usage (expired token)

Comment (kidroca) [reply src/libs/actions/Session/index.js:449]: This logic as well as the throttle might be unnecessary anymore
Depends on what we want to achieve 

Since `reauthenticate` would hook to the same authentication call it means that we won't be triggering multiple authentication calls anymore 

But the calls would still end up calling `Pusher.reconnect` and that might be undesired 

I prefer this logic over `throttle` as it guarantees a single call

Since this change (and the unit tests for it) are a bit out of scope I can move it  to a separate PR

File: src/libs/API.js
Comment (kidroca) [reply src/libs/API.js:79]: I'll push a change to rename this

**Review by kidroca — COMMENTED** [review]
Addressing open threads

File: src/libs/API.js
Comment (kidroca) [reply src/libs/API.js]: If there's a way to get rid of pausing, the change here would ease it, as everything related was moved here

The original code was pausing the queue - I didn't change that but cleaned it up and provided tests to confirm it still works as expected

File: src/libs/API.js
Comment (kidroca) [reply src/libs/API.js]: Renamed to `isRequestRetrieable`

File: src/libs/actions/Session/index.js
Comment (kidroca) [reply src/libs/actions/Session/index.js]: @johnmlee101 dropped this change from the PR
I've opened a new PR here: #6707

**Comment by kidroca** [comment]
Ready for another review

**Comment by mallenexpensify** [comment]
@johnmlee101 @roryabraham I see you're 'requested for review', 👀 plz

**Review by johnmlee101 — DISMISSED** [review]
With a limited knowledge set of this, it looks decent to me! I'll let @roryabraham do the final review

File: src/libs/API.js
Comment (marcaaron) [inline src/libs/API.js:191]: Correct me if wrong, but this here seems to be the main material change? 

If we get a network error (like a fetch error) now we just toss the request into the ether instead of retrying it?

Comment (marcaaron) [reply src/libs/API.js:191]: If that's correct. I'm sort of confused why we are doing this here instead of when we call `onError()` here

https://github.com/Expensify/App/blob/021ad5d5b057fcecd98ee70b6bf52938cca5fe32/src/libs/Network.js#L226-L228

Could we just add the request to the array of requests to retry instead?

https://github.com/Expensify/App/blob/021ad5d5b057fcecd98ee70b6bf52938cca5fe32/src/libs/Network.js#L243-L245

Seems easier to do that and then we won't need `onError()` to call `retryRequest()` which calls `Network.post()` etc.

**Review by marcaaron — CHANGES_REQUESTED** [review]
I haven't looked at the tests yet but did a deep dive on the changes. Overall, I'm hoping we can maybe scale things back a bit so we are not mixing together refactoring and addressing the main issue (which I think was to retry requests that failed because of some network condition).

File: src/libs/API.js
Comment (marcaaron) [reply src/libs/API.js:119]: TL;DR is the existing code here just pointless?

https://github.com/Expensify/App/blob/021ad5d5b057fcecd98ee70b6bf52938cca5fe32/src/libs/API.js#L112-L117

**Comment by marcaaron** [comment]
>This PR untangles some stuff and actually simplifies the netwotking code

Just want to add some thoughts here that there is a time to simplify and a time to get stuff done. When we choose to refactor complex code that our team has limited knowledge of it draws a significant amount of time and energy from those who do have that knowledge to review carefully.

@kidroca here's some feedback and some things that might be helpful in the future...

- try to follow our conventions to the best of your ability
- aim to do a minimum number of changes in PRs where the area of the code is particularly complex or not widely understood
- avoid fixing multiple issues simultaneously when working in a complex area

**Review by kidroca — COMMENTED** [review]
> > This PR untangles some stuff and actually simplifies the netwotking code
>
> Just want to add some thoughts here that there is a time to simplify and a time to get stuff done. When we choose to refactor complex code that our team has limited knowledge of it draws a significant amount of time and energy from those who do have that knowledge to review carefully.
> 
> @kidroca here's some feedback and some things that might be helpful in the future...

@marcaaron 
You making this sound like a bigger problem that it actually is
Why didn't you say so on your initial review?

It's a normal sized PR, considering half of it is automated tests it's actually small
Excuse me if it took you a significant amount of time review, I didn't think that it will!

I'll remove my refactors from this PR, but I wish you or anyone else had brought that up on one of the previous reviews

**Comment by kidroca** [comment]
Removed all the changes but the main thing, if you still want me to move the retry to Network and split error handling there leave a comment

**Comment by marcaaron** [comment]
>You making this sound like a bigger problem that it actually is

To be clear, my intention is only to give you some advice on how to grow and work better with us not to make a big deal or criticize. It's just feedback to keep in mind for the future.

>Why didn't you say so on your initial review?

This is my first review. I mentioned that it would [take me some time to catch up](https://github.com/Expensify/App/pull/6567#issuecomment-990214504) and that I'd be more comfortable with an initial refactor. I can see how that looks like a missed opportunity to be more direct initially, apologies for that, and thanks for the feedback!

Still, not sure it's productive to assign blame and not sure it would _change_ the feedback that I'm giving. This is a PR.  Author proposes changes. Reviewers exchange feedback on how to improve next time. Thanks for hearing me out!

File: src/libs/API.js
Comment (marcaaron) [reply src/libs/API.js:202]: >The opposite actually. When we get a network/fetch error we'll get right here where we retry if we're supposed to.

Yes, sorry I am describing the **current behavior** (without your changes)

I see your point about losing the logs. But we could also:

- Log that we are retrying because of a network error from within Network
- Add the request back into the queue right away 
- Not call the `onError()` because we don't need to reject the promise or call `setLoadingAndError()`

File: src/libs/API.js
Comment (marcaaron) [reply src/libs/API.js]: Cool. Maybe we can address this in a follow up issue if it becomes a problem. Not sure I am seeing the subtle bug yet but does sound like maybe persisted requests should not be saved with the `authToken` and we should always prefer the most recent one we have.

File: src/libs/API.js
Comment (marcaaron) [reply src/libs/API.js]: Yeah, maybe we can create a new issue for this as well? I'm not quite seeing the path for how this can happen or why we should be concerned about it - but guessing you would be able to document it and we can address it separately if it's concerning?

**Comment by marcaaron** [comment]
>if you still want me to move the retry to Network and split error handling there leave a comment

I do think we should try this, but could be convinced otherwise. It just seems it's easier to connect the dots from `HttpUtils.xhr()` back to the queue in the same `Network` lib vs. doing `onError() -> Network.registerErrorHandler() callback (API.js) -> Network.post()`.

File: src/libs/API.js
Comment (kidroca) [reply src/libs/API.js:202]: Sounds like a plan

Comment (kidroca) [reply src/libs/API.js:202]: @marcaaron This would actually not work

We're in async code here - pushing to `requestsToProcessOnNextRun` would do nothing

```js
.catch(error => {
  if (canRetryRequest(queuedRequest)) {
    requestsToProcessOnNextRun.push(queuedRequest); 
  }
});
```

Even if we make it work (by pushing directly to `networkRequestQueue`) we're introducing a delay of up to one second, until the network queue retries the request

The best thing seems to be to call `.post()` so that the request can be retried immediately

---

Maybe we can extract a `canRetryRequest` with just the logic so that it can be reused in Network and in the error handler in API 
And initiate the retry from the error handler

Instead of having code inline like 

```js
.catch(error => {
  if (canRetryRequest(queuedRequest)) {
    requestsToProcessOnNextRun.push(queuedRequest); 
  } else {
  // ...
});
```

I prefer to have it extracted to a function and the original code already does that

```js
.catch(error => onError(queuedRequest, error));
```

Comment (marcaaron) [reply src/libs/API.js:202]: Sorry, let's walk through a real example of a request that has fetch failures just to get on the same page? I might be missing the concern because the story in my head is that when fetch fails the network is likely inaccessible.

1. We go offline while the network queue is emptying
1. We get a fetch error
1. We put the request back into the queue 
1. We are offline now so the request queue won't be processed
4. We go online
5. The queue is processed
6. Request succeeds

What is the scenario where we have a fetch error and we would want to trigger an immediate retry?

Comment (marcaaron) [reply src/libs/API.js:202]: This POV makes sense to me, but I also wonder if we are pre-optimizing for an edge case where a fetch failure happens while we are online. If the request is failing for a reason do it just keep getting tried over and over again? Maybe the once per second will act as a throttle in that case and prevent an infinite network loop (which is a tradeoff when considering the artificial delay mentioned).

This also begs the question of whether we should even retry such things. They failed once while we were online - why should we expect them to succeed the next time? Maybe we should stop trying at some point or if a specific kind of error is caught?

Anyway, I'm finding it difficult to form an opinion without concrete examples of what these errors are. The important thing seems to be that the request is not lost entirely. I am not quite convinced it should be retried indefinitely and somewhat worried we may start to see infinite retry loops without some kind of system in place to prevent them.

Comment (kidroca) [reply src/libs/API.js:202]: Another reason to keep things in API is to avoid circular deps we don't have access to logging in Network.js
We can only use console.debug 

In API we can capture a retry and log about it

```js
    if (lodashGet(queuedRequest, 'data.shouldRetry')) {
        Log.hmmm('[API] Request is retriable, retrying...', queuedRequest);
        Network.post(queuedRequest.command, queuedRequest.data, queuedRequest.type, queuedRequest.shouldUseSecure)
```

And the error handler by default captures the fact that we've **handled** an error, while simply putting it back for retry does not

https://github.com/Expensify/App/blob/9c19841e5a1a738311142a3da49ffe74d9dfa20b/src/libs/API.js#L185-L186

Since we're already retrying requests on a few places inside API, wouldn't it be best to extract logic that can cover those as well ([handleExpiredAuthToken](https://github.com/Expensify/App/blob/9c19841e5a1a738311142a3da49ffe74d9dfa20b/src/libs/API.js#L100-L124)) (And all of those are hooked back to the [original requests' promise](https://github.com/Expensify/App/blob/9c19841e5a1a738311142a3da49ffe74d9dfa20b/src/libs/API.js#L168-L171))

Comment (marcaaron) [reply src/libs/API.js:202]: >Another reason to keep things in API is logging. To avoid circular deps we don't have access to logging in Network.js

That sounds reasonable to me in the short term because it is more work to solve - though this is one of the things we should eventually refactor. We should be able to log in the lib without worrying about a "circular dependency".

>And the error handler by default captures the fact that we've handled an error, while simply putting it back for retry does not

Yes, we would add new logs to say something else in this case like "there was an error and it is being retried"

>Since we're already retrying requests on a few places inside API, wouldn't it be best to extract logic that can cover those as well (handleExpiredAuthToken) (And all of those are hooked back to the original requests' promise)

Didn't really get this, but maybe do not understand. We are not losing the original promise context if we put it back into the request queue with the `resolve` and `reject` attached. If the request gets a response it will be handled.

Comment (kidroca) [reply src/libs/API.js:202]: > I am not quite convinced it should be retried indefinitely and somewhat worried we may start to see infinite retry loops without some kind of system in place to prevent them.

Yep, it's possible that a cors error can happen, or an error resulting from calling an url that no longer exists, but these are developer errors and would result in a deploy blocker anyway
- not sure if there's even a chance to call an `url` that doesn't exist - we're calling Expensify's api if an endpoint doesn't exist the server would respond with an error

The requests we retry here would eventually succeed 
But if you're interested in limiting retries to a finite number I have something here: https://github.com/Expensify/App/pull/6556

Comment (kidroca) [reply src/libs/API.js:202]: > > Since we're already retrying requests on a few places inside API, wouldn't it be best to extract logic that can cover those as well (handleExpiredAuthToken) (And all of those are hooked back to the original requests' promise)
> 
> Didn't really get this, but maybe do not understand. We are not losing the original promise context if we put it back into the request queue with the `resolve` and `reject` attached. If the request gets a response it will be handled.

I'm saying that we already have a `retry` logic inside API (calling Network.post)
And we're discussing whether we should add another `retry` logic in Network or API (to spare us from calling Network.post)
Let's keep using `Network.post` to retry the request like we do in the Response Handler (expired token)
It would still aid refactoring, when we identify exactly how it should happen

Comment (marcaaron) [reply src/libs/API.js:202]: >a cors error can happen, or an error resulting from calling an url that no longer exists
>these are developer errors and would result in a deploy blocker anyway

Are these the only possibilities we should expect to see? And if so and these should be deploy blocking events why would the proposed solution be to retry them? Should we maybe at least queue a log to alert in such cases in addition to retrying over and over?

>The requests we retry here would eventually succeed

Sorry, I'm not quite seeing where the eventual success happens.

>Let's keep using Network.post to retry the request like we do in the Response Handler (expired token)

As I mentioned already, I don't see a compelling reason to do it this way so it's not worth having this many back and forth exchanges to resolve.

Regardless, it seems more important to figure out how to handle the infinite retries concern. 

I'm not sure if there's a way to wrap that concern up quickly. But without a proposal to do so I'm not sure it's a great idea to retry these requests.

Comment (kidroca) [reply src/libs/API.js:202]: The reason to retry the request, besides everything else said
When we fail to reach the API for any reason - retry the request (if it's retriable) 

Keep the logic simple we can extend it if it becomes a problem

Comment (marcaaron) [reply src/libs/API.js:202]: >tell me what you need done

Sure, sorry for being indirect - that new plan sounds good for now

>- have the retry happen in Network.js
>- don't log anything
>- push the request directly to the queue

It would be good to log when the request happens so maybe next we can solve the circular dependency issues in a new PR.

**Comment by kidroca** [comment]
Moved logic to Network 
Ready for review

**Review by marcaaron — APPROVED** [review]

**Comment by mvtglobally** [comment]
@marcaaron  We are struggling a bit to get very slow Network where QA resources are. 1 tester was able to get low 4G from the basement ;), so it seem to be pass, but we are not able to run all environments properly.

**Comment by parasharrajat** [comment]
For controlling the speed on Desktop and Web. You can use https://nshipster.com/network-link-conditioner/ for macOS. @mvtglobally

**Comment by Jag96** [comment]
Tested on Desktop and looks good!

**Pull request merged**
